### DISCUSSION TITLE
Is it possible that there might be some unclosed HTTP streams in the MacOS implementation?

### DISCUSSION THREAD

**Discussion by Ansis100** [781flyingdutchman/background_downloader#396] [Q&A] [answered]
I didn't want to start an issue without concrete evidence.

We have a weird issue on MacOS where if the downloader has been uploading files for a long time (about 2.5k uploaded files), then the whole app will become buggy and unusable. We traced it down to an error in a different library (file_picker) mentioning "ProcessException: Too many open files". So the OS prohibits performing any additional file descriptor actions, hence the bugginess.

I was wondering if this might have been cause by background_downloader opening too many file or HTTP streams and not closing them. I ran `lsof -c app_process` and saw a new IPv4 file descriptor being created for every task that was processed by the downloader. That would be normal, but the descriptors were not disappearing from the list even after the upload was completed. I'm not 100% sure if this is caused by the package, but if it is, it could cause any app to eventually go into a state of not being able to open any further network connections.

**Comment by 781flyingdutchman** [comment]
Interesting and definitely possible - thanks for stress testing the plugin! I'll look through the code and see if this may indeed be the case - which plugin version are you using?

**Reply by Ansis100** [reply]
We're using the latest version - 8.5.6. 
Thank you for looking into this! :)

**Reply by Ansis100** [reply]
In my testing I noticed that all the old connections are CLOSE_WAIT, which supposedly means that [something is wrong in the first place.](https://stackoverflow.com/a/26928451/7858553)
I'm not sure on the specifics of UNIX sockets, but I'm guessing somewhere, something is needlessly opening a new connection for every upload even though it should be reusing the old connection?

**Reply by Ansis100** [reply]
I think I found the issue. In my case it was the `http.StreamedRequest` opening connections in `binaryUpload`, finishing them with a CLOSE_WAIT, but not removing them.

https://github.com/781flyingdutchman/background_downloader/blob/3ec71d67337c1a6d3002c6c218f8a007f8726a96/lib/src/desktop/upload_isolate.dart#L49-L56

After adding a `request.persistentConnection = false;` after line 53, the CLOSE_WAIT descriptors seem to disappear.

I haven't yet tested if it affects the upload connections in any other way. The `multipartUpload` also uses `StreamedRequest`, I haven't tested that either.

**Reply by 781flyingdutchman** [reply]
Thanks for looking into this thoroughly. I'll mark this as a bug - do you want to submit a PR for this? You'd have to make sure all integration tests in the example app pass (especially for the upload ones) for all platforms (Windows, MacOS, Linux in particular, given this affects the desktop platforms only) and would indeed suggest doing this also for the multi-part upload.  You shoudl base your PR on the `dev` branch as there is already some other stuff being tested there.  If not, let me know and I will add it when I get to it.

**Reply by 781flyingdutchman** [reply]
I looked into this a bit and it seems that it is preferred to set persistentConnection to true, so the client can reuse a connection to the same server. However, the isolate architecture for desktop downloads and uploads forces the creation of a new httpClient for each task, which is likely why you're seeing connections all over the place. It would be great if you could test if setting it to false leads to slower performance if you upload or download a large number of smaller files to the same server. I don't think so, so suggest we change to false for all uploads and downloads on desktop.

**Reply by 781flyingdutchman** [reply]
In my very basic testing I saw no adverse effect of setting persistentConnection to false, and of closing the HTTP client when the isolate exists. Both are likely better from a memory and performance perspective, so adding this to the next release.

**Comment by 781flyingdutchman** [comment, marked as answer]
Implemented in V8.6.0

**Reply by Ansis100** [reply]
Thank you! 
I'll keep an eye on the performance, but our testers didn't report any noticeable difference.
